### ISSUE TITLE
Existing system language survey

### ISSUE THREAD

**Issue by arBmind** [rebuild-lang/rfcs#2]
A list of acknowledgable system programming languages.

We should not take features that don't fit our goals. But it's always good to take terminology and inspirations.

## C

http://www.open-std.org/jtc1/sc22/wg14/

very mature system programming language.

- [x] the PC world is build on C
- [x] well defined ABI and calling conventions
- [x] no required std library (only what you use)
- [ ] all burden for resource management on developers
- [ ] only precompiler macros
- [ ] depends on precompiler "#include"-system
- [ ] no standard for error handling
- [ ] no std for higher level concurrent concepts

## C++

https://isocpp.org/

modern system development language

- [x] native OOP support
- [x] large scale of abstractions
- [x] code and type generation with templates
- [x] RAII resource management
- [x] exceptions
- [ ] depends on precompiler "#include"-system
- [ ] tends to slow compile times
- [ ] templates use verbose syntax and have limitations
- [ ] no defined ABI
- [ ] requires a std-library runtime for many features

## D

https://dlang.org/

language to replace C

- [x] import system
- [x] RAII & scopes
- [x] interface & class OOP
- [x] explicit compile time execution
- [x] module/import system
- [ ] garbage collector (not mandatory but almost unusable without)
- [ ] not many libraries

## Go

https://golang.org/

server programming language

- [x] no external runtime required
- [x] package/import system
- [x] omitable parenthesis & semicolons
- [x] channel based goroutines (allows stacked coroutine)
- [ ] garbage collector
- [ ] no scopes for member function

## Pony

http://www.ponylang.org/

actor-model, capabilities-secure, high performance programming language

- [x] directory based package/use system
- [x] language based actors
- [x] actor based garbage collector (faster than other options)
- [x] full soundness check on compile time
- [x] type expressions
- [ ] actor based garbage collector (not zero abstraction)
- [ ] only simple generics
- [ ] not very mature (0.2.1)

## Crystal

http://crystal-lang.org/

ruby inspired syntax with static typing & compile-time evaluation

- [x] explicit, easy to read syntax
- [x] ast based templating macro system (mustache syntax)
- [x] fiber & channel based actor concurrency
- [ ] require system (basically #include)
- [ ] garbage collector
- [ ] no parallelism so far
- [ ] still alpha

## Rust

https://www.rust-lang.org/

systems programming language with compile time safety

- [x] sane defaults (const vs. mut, ...)
- [x] guaranteed memory safety (borrow system)
- [x] ast based pattern matching macro system
- [x] module/use system with explicit publishing
- [x] package manager crates.io included
- [x] no garbage collector
- [ ] limited generics (trade off for safety features without long compile times)
- [ ] very strict compiler (stdlib can do stuff users cannot)
- [ ] hard to learn for classical OOP programmers
- [ ] not all libraries

## Swift

https://developer.apple.com/swift/

modern system programming language from Apple to superseed C and Objective C

- [x] module/import system
- [ ] Taylored to Apple platforms
- [ ] ARC for memory management
- [ ] Limited generics system
- [ ] Language wired features (properties, protocols, weak, ...)
- [ ] No macro system (Only conditional compiling)

## Felix

http://felix-lang.org/

High performance scripting language.

- [x] differentiates fun, proc & gen methods
- [x] coroutines
- [ ] garbage collector
- [ ] very magically (all built in)

## Sylph

https://eev.ee/blog/2015/02/28/sylph-the-programming-language-i-want/

a collection of thoughts on a better language.

- [ ] non existing (only a blog post of wishes and opinions)

## Jai

https://www.youtube.com/playlist?list=PLmV5I2fxaiCKfxMBrNsU1kgKJXD3PkyxO

Jonathan Blows game programming language.

- [x] Structure of Array support
- [x] Run any code as compile time script
- [x] Imperative meta programming with compiler events and AST
- [x] In code build system
- [ ] no public code (only Youtube videos)
- [ ] hard aim at game development

---
Basically there remains much to improve on every aspect of system programming languages.

If you have more please drop a comment.

**Comment by liigo** [comment]
I don't think Rust's stdlib can do stuff users cannot. Could you elaborate? I known that its stdlib do use some lang_items (defined by compiler), which user's code can use too.

**Comment by arBmind** [comment]
@liigo Some details to support my findings. I don't think it's the most annoying one.

Take a look at all the `#![feature]` calls in https://github.com/rust-lang/rust/blob/master/src/libstd/lib.rs#L213 These are only allowed in nightly compilers for normal users.

The smallest example I came up to support this.
Wrap a lambda expression as a function into a box.
The following code works in nightly. I struggled to find a stable solution.

```rust
#![feature(box_syntax)]
use std::ops::{Add};

fn factory<'a, T, Q>(lhs: T) -> Box<Fn(Q) -> <T as Add<Q>>::Output + 'a> where T : 'a + Add<Q> + Copy
{
    box move |rhs : Q| lhs + rhs
}

fn dummy() -> Box<Fn(i32) -> i32> { factory(4) }

fn main() {
    let answer = dummy()(2);
    assert_eq!(6 , answer);
    println!("{}", answer);
}
```

I don't claim that no stable solution exists. The issue is simple that by reading the standard library you learn syntax that you are not supposed to use. In C++ reading and understanding the STL gives you a very good feeling how to use the even most advanced C++  language features.

On the other hand the syntax for this simple example is already quite complex. It is not very teaching friendly.

**Comment by arBmind** [comment]
I moved the existing languages to the wiki and added zig [Existing-system-language-survey](https://github.com/rebuild-lang/rfcs/wiki/Existing-system-language-survey)

**Comment by liigo** [comment]
Thanks for reply. In Rust, users are prevent from using some *unstable* features with *stable* compiler. (I currently can not think of a soundness reason to allow user doing that - it's full of tricks.) But they are still allowed to use them with nightly compiler, i.e. unstable compiler, and so any user *can* create his own stdlib from ground, in theory. The stdlib is nothing special in this case.

**Comment by liigo** [comment]
As an example, [serde](https://github.com/serde-rs/serde/), a third-party library of Rust, do use unstable features in https://github.com/serde-rs/serde/blob/master/serde/src/lib.rs#L15, in the same way as stdlib do.

**Comment by arBmind** [comment]
@liigo Thanks for defending Rust. I know it's a design decision, that cannot be easily overcome. All I am saying is that this removes the learning opportunity from the standard library sources. For everything else it works.

My suggestion for Rebuild Lang is that the standard library is written in the stable language. By enabling the users to grow the language, the features of the compiler should be quite stable.

**Comment by liigo** [comment]
> My suggestion for Rebuild Lang is that the standard library is written in the stable language.

You can not even start your work *before* your language is stable.

**Comment by arBmind** [comment]
> You can not even start your work before your language is stable.

That is right. I am currently collecting ideas. One of my thoughts is to make the core language as simple as possible. And keep almost everything implementable by the user. My hope is that the core is small and will be stabilized very early on. This should lead to a market of competing ideas, where the best ones, will make it to the standard library.

**Comment by liigo** [comment]
That will be a great work, I'm eager to your Rebuild, a future maybe-perfect systems language. Good luck, @arBmind .

> to make the core language as simple as possible

Rust language separates features as stable and unstable.

> keep almost everything implementable by the user.

I'd like to say, Rust's `std` and `core` library are both implemented by the user, in some degree. Yes, users can make stable libraries by using unstable compiler and unstable language features. Just wrap something up as implementation details.

> where the best ones, will make it to the standard library

This is what Rust already done (or doing).
